Update Trend chart copy for special cases (on frontend)

shaunlebron opened the pull request:
Part of [Trend Chart improvements epic](https://github.com/metabase/metabase/issues/33411)

Will close (after feature branch merged): https://github.com/metabase/metabase/issues/8919
Will close (after feature branch merged): https://github.com/metabase/metabase/issues/23385

## User-facing Changes

https://github.com/metabase/metabase/assets/22608765/89ab247f-8404-40e0-9856-d4c64ccdf831

This PR adds the following updates:

* corrections to our copy for special cases:
  * **previous value is 0**: display infinite percents with `↑ ∞% · vs. previous month: 0`
  * **previous value is the same as current value**: handle 0% change with `No change · vs. previous month`
  * **previous period is a null value**: skip rows with null values when finding the previous value to compare against
    * **no previous values to compare**: handle missing data case with `N/A · vs. previous month: (empty)`
  * truncating `prevDate` based on `nextDate`:
    * if `prevDate` and `nextDate` are on the same day, remove the day, month, and year from the comparison period string
    * if `prevDate` and `nextDate` are in the same year, remove the year from the comparison period string
* We also fix dark mode, by using opacity to achieve same styles in both modes:

before | after
---|---
![CleanShot 2023-11-07 at 10 44 11@2x](https://github.com/metabase/metabase/assets/116838/9eb5f087-f82d-44ed-96e0-e593cb92efd1) | ![CleanShot 2023-11-07 at 10 43 40@2x](https://github.com/metabase/metabase/assets/116838/ad854a34-6033-41d1-a56d-dce5c1b5e853)
![CleanShot 2023-11-07 at 10 46 59@2x](https://github.com/metabase/metabase/assets/116838/b17eb9a5-6c95-4d8e-84d4-a4a8977f0f6e) | ![CleanShot 2023-11-07 at 10 47 29@2x](https://github.com/metabase/metabase/assets/116838/7b5f8cdd-b391-440d-8a1b-e9f865941a3d)

## Static viz is not updated yet

Static viz will *not* show the aforementioned changes, but there is now a `computeTrend` function used on the frontend which can be used by the backend static viz renderer when wired up correctly in a future PR.

## How to review

SmartScalar had to be re-organized for the requirements of this epic, so I apologize for the large diff— but I hope you will find it more readable than our previous version.

To start, I recommend first clicking the files listed below and perusing them on their own, since they are largely new.  Then reference my diff comments to compare to their original implementations, and check the unit tests.

* [SmartScalar/compute.js](https://github.com/metabase/metabase/blob/c4e600a3015c4d3bc0b3737643476678e7fef400/frontend/src/metabase/visualizations/visualizations/SmartScalar/compute.js) is a new file that has everything needed to compute the trend data, and also some formatted display strings to be shared between the UI and static viz (in the `display` properties).
* [SmartScalar/SmartScalar.js](https://github.com/metabase/metabase/blob/c4e600a3015c4d3bc0b3737643476678e7fef400/frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx) is the main component which had a lot of logic moved into either `computeTrend` or `PreviousValueComponent`

shaunlebron committed:
add computeTrend

iethree commented:
This is a great list of independent tasks here:

> - fix dark mode
> - display infinite percents correctly
> - handle missing data case with special message
> - handle 0% change with special message
> - when selecting previous value to compare, find the most recent row with a non-null value (instead of just the previous row)
> - prepare data for rendering to static viz

Would it maybe make sense to open each of these as a separate little PR? that way we could get some quick easy wins into master one at a time? Would also make it much easier to review each one. 🙏

shaunlebron commented:
@iethree I would if I could.  It’s a bigger refactor, so the individual changes are 2-3 lines each on top of the refactoring.

shaunlebron committed:
fix unit test

shaunlebron committed:
fix e2e test

shaunlebron commented on e2e/test/scenarios/visualizations/smartscalar-trend.cy.spec.js:
```diff
@@ -29,7 +29,5 @@ describe("scenarios > visualizations > scalar", () => {
 
     cy.get(".ScalarValue").contains("100");
-    // eslint-disable-next-line no-unscoped-text-selectors -- deprecated usage
-    cy.findByText("Nothing to compare for the previous month.");
```
I removed this check because it’s only necessary to confirm that this visualization displays a value, not what the comparison is.  See #13710.

_Extra context— Our new frontend `computeTrend` function skips over null gaps until finding a valid month, and changes this message.  I opted not to correct this message because the rows are not sorted by date here, which causes the backend trend calculation to trip up since the previous row is not exactly one month adjacent to the source row.  The frontend calculation now displays the previous row regardless of what it is._

shaunlebron commented on frontend/src/metabase/css/dashboard.css:
```diff
@@ -263,24 +263,6 @@
-    display: inherit;
-  }
-}
-
```
I removed this because we have a new system for responsively toggling this value when it doesn’t fit in the card.

_Extra context— the id `#SmartScalar-PreviousValue` is only used here as a hack to hide long numbers at the end of `↑ 50% · vs. previous month: ______` on automated dashboards (via x-ray) when the page isn’t wide enough, but this is a hack because it doesn’t measure the card width, which our new system does instead._

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.ts:
```diff
@@ -73,42 +63,3 @@ export const getValueHeight = (height: number): number => {
-    isPreviousValueTruncated: matchIdx !== 0,
-    fittedPreviousValue: previousValueCandidates[matchIdx],
-  };
-};
```
Everything this function was doing is now inlined in a shorter way here:
https://github.com/metabase/metabase/pull/35184/files#diff-b0fbfd8b0eb7fcb57feb5b0b2df90ad1925947bac9fc8aefab49fbd15dfbc76aR106-R128

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.ts:
```diff
@@ -21,36 +20,27 @@ export const formatChangeAutoPrecision = (
+          family: fontFamily,
+          weight: fontWeight,
+        }).width <= width,
+    ) ?? formatChange(change, { maximumFractionDigits: 0 });
```
a functional refactoring for the next guy who has to look at it.  [existing unit tests](https://github.com/metabase/metabase/blob/67a1d7847390308e52d0f1cc83fbd66eeac8724b/frontend/src/metabase/visualizations/visualizations/SmartScalar/utils.unit.spec.ts#L49-L74) verify that it still works.

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/compute.unit.spec.js:
```diff
@@ -0,0 +1,140 @@
+        ],
+        expected: "100; Nov 2019; ↓ 66.67% vs. Sep 2019: 300",
+      },
+    ])("$description", ({ rows, expected }) => {
```
Start here to understand how our new `computeTrend` function returns everything needed to display our trend chart, both on the UI and the static viz.

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/compute.js:
```diff
@@ -0,0 +1,184 @@
+
+export const PREVIOUS_VALUE_MISSING = "PREVIOUS_VALUE_MISSING";
+export const PREVIOUS_VALUE_SAME = "PREVIOUS_VALUE_SAME";
+export const PREVIOUS_VALUE_CHANGED = "PREVIOUS_VALUE_CHANGED";
```
Enumerating of the different cases that have to be displayed differently.

shaunlebron committed:
explicitly fallback to day unit if backend doesn’t provide one

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/compute.unit.spec.js:
```diff
@@ -0,0 +1,156 @@
+        dateUnit: null,
+        expected: "300; Nov 1, 2019; ↑ 200% vs. Sep 1, 2019: 100",
+      },
+    ])("$description", ({ rows, expected, dateUnit }) => {
```
Start here to understand how our new `computeTrend` function returns everything needed to display our trend chart, both on the UI and the static viz.

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/compute.js:
```diff
@@ -0,0 +1,185 @@
+const FALLBACK_DATE_UNIT = "day";
+
+// compute the percent change between two values (a → b)
+export function computeChange(a, b) {
```
ported from clojure: https://github.com/metabase/metabase/blob/d2ecd17/src/metabase/sync/analyze/fingerprint/insights.clj#L32-L42

For readability, I split up the cases into comments, which are validated in [unit tests below](https://github.com/metabase/metabase/pull/35184/files#diff-5319611e9f4c7a99acd8792fbeb80a7961e52432a545beaad54ab259db37c4b2R8-R59).

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.unit.spec.js:
```diff
@@ -35,20 +35,8 @@ describe("SmartScalar", () => {
-        "previous-value": 200,
-        unit: "month",
-        col: "Sum",
-      },
```
These insight values are now computed in the frontend automatically, so we remove them in these unit tests to ensure we’re using those instead.

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:
```diff
@@ -71,125 +66,131 @@ const ScalarPeriod = ({ lines = 2, period, onClick }) => (
-        compact: true,
-      })}
-    </PreviousValueLabel>
-  );
```
I did my best to unify a lot of our elements between the card and tooltip to deduplicate this kind of work.  All work related to comparison display is cleaned up and moved  to `PreviousValueComparison`.

shaunlebron committed:
fix empty data cases

JesseSDevaney commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.styled.tsx:
```diff
@@ -2,11 +2,18 @@ import styled from "@emotion/styled";
 export const Variation = styled.div`
   color: ${props => props.color};
+  opacity: ${props =>
+    props.color == null ? detailsNumberOpacity * detailsOpacity : 1};
```
```javascript
import { isEmpty } from "metabase/lib/validate";
```

```suggestion
    isEmpty(props.color) ? detailsNumberOpacity * detailsOpacity : 1};
```

JesseSDevaney commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.styled.tsx:
```diff
@@ -2,11 +2,18 @@ import styled from "@emotion/styled";
-import { color } from "metabase/lib/colors";
 import { space } from "metabase/styled-components/theme";
 
+// NIGHT-MODE TEXT HACK:
+// Our "fullscreen-night-text" className only supports one shade of text (white).
+// As a temporary fix, we create dark-mode analogs for "text-medium" and "text-light" using opacity:
+const detailsOpacity = 0.85; // approximates color("text-medium")
+const detailsNumberOpacity = 0.7; // approximates color("text-light")
+
```
I do not quite understand why we are using opacity instead of directly setting the color here. Can you explain it to me?

JesseSDevaney commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:
```diff
@@ -1,62 +1,57 @@
 /* eslint-disable react/prop-types */
```
Can we convert this into TypeScript or is there a reason this file was not converted into TypeScript?

JesseSDevaney commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:
```diff
@@ -71,125 +66,131 @@ const ScalarPeriod = ({ lines = 2, period, onClick }) => (
   totalNumGridCols,
   fontFamily,
 }) {
+  const fontSize = "0.875rem";
```
Why are we hard-coding `fontSize` in this component?

JesseSDevaney reviewed (changes requested):
Hey Shaun - thanks for working on this and getting this 99% of the way there! :bowing_man:  

There are some things I think should be addressed before merging. Most of my comments are nits, some are about correctness, and some are about educating me. 

I am hoping we can find a time to pair on this Monday to go over the changes and work-out what should be changed and what does not need to be before we merge. Until then, I hope you have a lovely weekend :palm_tree:

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/compute.js:
```diff
@@ -0,0 +1,186 @@
+  const title =
+    date == null || datesAreContinuous
+      ? t`previous ${dateUnitDisplay}`
+      : formatDateTimeRangeWithUnit([date], dateUnit, { compact: true }); // FIXME: elide part of the prevDate in common with lastDate
```
The figma design shows that the date here should omit the year:
<img width="545" alt="CleanShot 2023-11-20 at 10 15 05@2x" src="https://github.com/metabase/metabase/assets/116838/cbf64f98-ed43-4a2b-894b-0709d11e5889">

JesseSDevaney replied:
See commit [7ffa506](https://github.com/metabase/metabase/pull/35184/commits/7ffa506e28ae57f88a9eb11a831c26ba0f04c275)

JesseSDevaney committed:
rename function parameters

JesseSDevaney committed:
correct and simplify computeChange function

- when prevVal < 0 and nextVal < 0, it would return the wrong percentChange. (nextVal - prevVal) / Math.abs(prevVal) fixes this issue.
  - i.e. prevVal = -5, nextVal = -10 -> computeChange(-5, -10) -> computeChange(--10, --5) -> -50% but we should it expect to be -100% since it dropped by a factor of Math.abs(prevVal)

JesseSDevaney committed:
refactor PREVIOUS_VALUE options

JesseSDevaney committed:
correct and simplify computeChange further

(-) -> (0) should be + increase (our end case already accounts for this behavior)

JesseSDevaney committed:
update unit tests to reflect correctness changes to computeChange

JesseSDevaney committed:
refactor for readability

JesseSDevaney committed:
refactor to use isEmpty instead of ==/!= null

JesseSDevaney committed:
refactor variable names for readability

JesseSDevaney committed:
refactor nested ternaries into function

JesseSDevaney committed:
refactor for readability

JesseSDevaney committed:
refactor variable names

JesseSDevaney committed:
re-use end of array index variable

JesseSDevaney committed:
re-word comments

JesseSDevaney committed:
refactor consumers based on return variable re-naming

JesseSDevaney committed:
add missing test cases

JesseSDevaney committed:
nit refactor

JesseSDevaney committed:
Merge branch 'master' into frontend-trend-calc

JesseSDevaney committed:
re-arrange functions for readability

JesseSDevaney committed:
truncate prevDate based on nextDate

- if prevDate is same day as nextDate, remove the day, month, and year from the comparisonPeriodStr
- if prevDate is same year as nextDate, remove the year from the comparisonPeriodStr

JesseSDevaney committed:
fix comment

shaunlebron commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/compute.unit.spec.js:
```diff
@@ -125,11 +125,11 @@ describe("SmartScalar > compute", () => {
         description:
           "should correctly display percent decrease over missing row",
```
The description for the test case doesn’t make sense anymore after your change below.

This “missing row” here is intended to be October.  If the “latest” data is for November, and the most recent data is for September, then the comparison would read “Sep” instead of “previous month”.  Does that make sense?

JesseSDevaney replied:
Yes, thanks for catching that!

JesseSDevaney committed:
fix test

cdeweyx reviewed (approved):
Works as expected 👍

alxnddr commented on frontend/src/metabase/lib/formatting/date.tsx:
```diff
@@ -634,6 +630,29 @@ export function formatDateTimeRangeWithUnit(
+    const dayRegex = /[\s\S]+,*\s*YYYY,*\s*/;
+
+    // day display is hide-able (if date being compared to is same day)
+    if (options.removeDay) {
```
The new `removeDay` and  `removeYear` options assume `formatStr` date format and probably won't work on the other date formats. 

I would assume that instead of replacing formatStr we need to compose it right away without a day or a year before it gets to this function

alxnddr replied:
It would be also nice to add more unit test son `date.tsx` specifically to verify more combinations on different formats since it is our common formatting code and it should have good coverage

alxnddr commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.jsx:
```diff
@@ -71,16 +66,118 @@ const ScalarPeriod = ({ lines = 2, period, onClick }) => (
+    display,
+  } = comparison;
+
+  const arrowIconName = { "↓": "arrow_down", "↑": "arrow_up" }[changeArrow];
```
I'd avoid using symbols like `↓` that are not convenient to type and instead just used `arrow_down`. Seems like these symbols themselves are not used anywhere in UI

alxnddr commented on frontend/src/metabase/visualizations/visualizations/SmartScalar/SmartScalar.styled.tsx:
```diff
@@ -2,11 +2,19 @@ import styled from "@emotion/styled";
+import { isEmpty } from "metabase/lib/validate";
+
+// NIGHT-MODE TEXT HACK:
+// Our "fullscreen-night-text" className only supports one shade of text (white).
```
Reimplementing night mode in a more convenient way for our styles is definitely out of the scope of this project.

Can we still avoid such hacks and use nested styles?
```
  .fullscreen-night-text & {
    color: ${...};
  }
```

alxnddr reviewed (commented):
Great work! I'm happy to see so many added specs on SmartScalar

I left a few comments, can you please have a look?

JesseSDevaney committed:
instead of ↓ and ↑ symbols, use icon names

JesseSDevaney committed:
fix night mode hack

JesseSDevaney committed:
refactor formatDateTimeRangeWithUnit behavior

- refactor how options.removeDay and options.removeYear work
- reorganize the layout to make the conditional branches easier to see

JesseSDevaney committed:
fix tooltip styles

JesseSDevaney committed:
Add tests to cover new options

- options.removeDay
- options.removeYear

JesseSDevaney committed:
add option descriptors to test string

JesseSDevaney committed:
Merge branch 'master' into frontend-trend-calc
